Objective-C/CBLDatabase.mm

```diff
@@ -722,45 +722,92 @@ - (nullable CBLScope*) scopeWithName: (NSString*)name error: (NSError**)error {
+    }
+    FLArray_Release(list);
+    
+    return [NSArray arrayWithArray: collections];
 }
 
-- (nullable CBLCollection*) createCollectionWithName: (NSString*)name
-                                      scope: (nullable NSString*)scope
-                                      error: (NSError**)error {
+- (nullable CBLCollection*) createCollectionWithName: (NSString*)collectionName
+                                               scope: (nullable NSString*)scopeName
+                                               error: (NSError**)error {
+    scopeName = scopeName.length > 0 ? scopeName : kCBLDefaultScopeName;
+    CBLStringBytes cName(collectionName);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
     
-    // TODO: add implementation
+    C4Collection* c4collection;
+    CBL_LOCK(self) {
+        [self mustBeOpen];
+        
+        C4Error c4err = {};
+        c4collection = c4db_createCollection(_c4db, spec, &c4err);
+        if (!c4collection) {
+            // TODO: handle error
+            return nil;
+        }
+    }
     
-    CBLScope* s = [[CBLScope alloc] initWithName: scope ?: kCBLDefaultScopeName error: error];
-    return [[CBLCollection alloc] initWithName: name scope: s error: error];
+    return [[CBLCollection alloc] initWithDB: self c4collection: c4collection error: error];
 }
 
 - (nullable CBLCollection*) collectionWithName: (NSString*)name
-                                scope: (nullable NSString*)scope
-                                error: (NSError**)error {
+                                         scope: (nullable NSString*)scope
+                                         error: (NSError**)error {
     
-    // TODO: add implementation
+    // TODO: Its expensive to get c4collection everytime from LC
+    
+    NSString* scopeName = scope ?: kCBLDefaultScopeName;
+    CBLStringBytes cName(name);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
     
-    CBLScope* s = [[CBLScope alloc] initWithName: scope ?: kCBLDefaultScopeName error: nil];
-    return [[CBLCollection alloc] initWithName: name scope: s error: nil];
+    __block C4Collection* c;
+    CBL_LOCK(self) {
+        [self mustBeOpen];
+        
+        c = c4db_getCollection(_c4db, spec);
+        if (!c) {
+            // TODO: handle error
+            return nil;
+        }
+    }
+    
+    return [[CBLCollection alloc] initWithDB: self c4collection: c error: error];
 }
 
 - (BOOL) deleteCollectionWithName: (NSString*)name
                             scope: (nullable NSString*)scope
                             error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    NSString* scopeName = scope ?: kCBLDefaultScopeName;
+    CBLStringBytes cName(name);
+    CBLStringBytes sName(scopeName);
+    C4CollectionSpec spec = { .name = cName, .scope = sName };
+    CBL_LOCK(self) {
+        C4Error c4err = {};
```

Comment:
Need to check if the collection is valid before doing anything as well.

Reply:
does this mean, we have to get the c4collection from collectionSpec and then check validity and then delete? which is not very efficient right? instead LC can do the check, whether it exists, if not throw error? 

Or can we guarantee, getting a collection means, it exists and is valid. No need to further check for isValid?

Reply:
I see `c4db_hasCollection` API